R/read_file.R

```diff
@@ -0,0 +1,80 @@
+#' Read files with extension .fst, .csv, and .rda (.rds and .RData)
+#'
+#' @param file_name Character; path to a file storing a single object.
+#'
+#' @noRd
+#' @examples
+#' # Odd case: .RData saved with extension .rda
+#' path <- tempfile(fileext = ".rda")
+#'
+#' data1 <- data.frame(x = 1)
+#' saveRDS(data1, path)
+#' read_file(path)
+#'
+#' # More natural case: .RData saved as .rda
+#' data2 <- data.frame(x = 1)
+#' save(data2, file = path)
+#' read_file(path)
+#'
+#' # Fails by design: .RData with more than 1 object
+#'
+#' save(data1, data2, file = path)
+#' # Fails
+#' try(read_file(path))
+read_file <- function(path) {
+  # FIXME: A warning seems like a bad idea. It should likely be an error
+  if (!fs::file_exists(path)) {
+    rlang::warn("The `path` does not exist: {path}")
+    return(invisible(path))
+  }
+
+  abort_unsupported_extension(path)
+
+  extension <- fs::path_ext(path)
+  switch(
+    extension,
+    "rda" = read_rda(path),
+    "fst" = fst::read_fst(path),
+    "csv" = readr::read_csv(path),
+    rlang::abort("Extension is supported but something unexpected happened.")
+  )
+}
+
+abort_unsupported_extension <- function(path) {
+  if (!supported_extension(path)) {
+    rlang::abort(glue::glue("Unsupported `path` extension: {path}"))
+  }
+
+  invisible(path)
+}
+
+supported_extension <- function(path) {
+  fs::path_ext(path) %in% c("rda", "csv", "fst")
+}
```

Comment:
I don't understand/agree with the idea of creating single line functions that are only ever called once. In my opinion, it substantially degrades the readability of the code. I would prefer this type of code to be inline with the code that calls it so that it's much easier to follow.

Reply:
I'm influenced by "Clean code" by Robert C. Martin. The name of the function acts as a comment to explain what the function does, but hides the implementation details to make it easier to read and easier to maintain. The high-level function does not need to know the implementation details of the lower level functions; this ensures that when the low level details change, the high level functions remain the same.

If the function does what its name promises, the reader can understand it from the overview that the high level function provides. When the reader wants to know the details, they can jump to the definition of the function. The code reads like a magazine article, with an overview at the top (high level function) and details becoming finer grained as you descend.